MCO-1283: adds OCL quickstart and troubleshooting guides (#4544)
openshift/machine-config-operator: docs/onclusterlayering-quickstart.md

### Initial Setup

First, we'll create the necessary objects that we'll need in advance:

```bash
#!/usr/bin/env bash

# (required): Clone the global image pull secret into a secret called
# "global-pull-secret-copy" within the MCO namespace. This will be used to pull
# the base OS image:
oc create secret docker-registry global-pull-secret-copy \
  --namespace "openshift-machine-config-operator" \
  --from-file=.dockerconfigjson=<(oc get secret/pull-secret -n openshift-config -o go-template='{{index .data ".dockerconfigjson" | base64decode}}')

# (optional): Create an ImageStream which will act as our image registry. This
# step can be omitted if you are using an external image registry such as
# Quay.io, though you will need to create the push and pull secrets in Quay.io
# as well as inside your cluster.
oc create imagestream os-images -n openshift-machine-config-operator

# (optional): To get the image registry pullspec for your ImageStream, you can run the following command:
# This will return "image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/os-images"

> oc get imagestrea/os-images  -n openshift-machine-config-operator -o=jsonpath='{.status.dockerImageRepository}'

Author: Done.

      # You'll note that you need to provide an architecture. This is because
      # this will eventually support multiarch clusters.
      - containerfileArch: noarch
        content: |-
          FROM quay.io/centos/centos:stream9 AS centos
          RUN dnf install -y epel-release

          FROM configs AS final
          COPY --from=centos /etc/yum.repos.d /etc/yum.repos.d
          COPY --from=centos /etc/pki/rpm-gpg/RPM-GPG-KEY-* /etc/pki/rpm-gpg/
          RUN sed -i 's/\$stream/9-stream/g' /etc/yum.repos.d/centos*.repo && \
              rpm-ostree install cowsay && \
              ostree container commit
    # Here is where you can select an image builder type. For now, we only
    # support a pod image builder that we maintain ourselves. Future
    # integrations can / will include other build system integrations.
    imageBuilder:
      imageBuilderType: PodImageBuilder
    # The Machine OS Builder needs to know what pull secret it can use to pull
    # the base OS image.
    baseImagePullSecret:
      name: global-pull-secret-copy
    # Here is where you specify the name of the push secret you use to push
    # your newly-built image to.
    renderedImagePushSecret:

>       name: builder-dockercfg-123

Author: Within this context, I think it can be left as-is because this represents the output of a script as opposed to a generic example where someone may be asked to fill in the blanks.

    # Here is where you specify the image registry to push your newly-built
    # images to. In this example, we're using an ImageStream, but one can
    # easily use an image registry such as Quay.io.
    renderedImagePushspec: image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/os-image:latest
  buildOutputs:
    # Here is where you specify what image will be used on all of your nodes to
    # pull the newly-built image.
    currentImagePullSecret:

>       name: builder-dockercfg-123

Author: Within this context, I think it can be left as-is because this represents the output of a script as opposed to a generic example where someone may be asked to fill in the blanks.

```

Now that we've done this, we can apply the layered MachineOSConfig to our cluster:

```console

> $ oc apply -f ./layered-machineosconfig.yaml

Author: It's not needed. It was just something left over from when I was experimenting with Jupyter.

    --for=jsonpath='{.metadata.annotations.machineconfiguration\.openshift\.io/state}=Done' \
    "node/$nodeName"
```

Once the node has finished the update, we can interrogate the node state like we did before:

```console
$ oc debug node/ip-10-0-18-226.ec2.internal -- chroot /host rpm-ostree status

Starting pod/ip-10-0-18-226ec2internal-debug-xvbx4 ...
To use host binaries, run `chroot /host`

State: idle
Deployments:
* ostree-unverified-registry:image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/os-images@sha256:c47856f56e1fdb7c9d10a1658e4ea85fbea44d71fb0e82898d152b47e0f894c6
                    Digest: sha256:c47856f56e1fdb7c9d10a1658e4ea85fbea44d71fb0e82898d152b47e0f894c6
                  Version: 417.94.202408190141-0 (2024-08-20T14:40:42Z)

Removing debug pod ...
```

We can see now that our node has booted into our new OS image. But does it have `cowsay` installed?

```console
$ oc debug node/ip-10-0-18-226.ec2.internal -- chroot /host rpm-ostree status

Starting pod/ip-10-0-18-226ec2internal-debug-6vvvf ...
To use host binaries, run `chroot /host`

  ______________________ 

> < Hello layered world! >

Author: I had to change this to use `tree` instead of `cowsay`.

  ---------------------- 
        \   ^__^
          \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

Removing debug pod ...
```

## Conclusion

> At this point, we now have a customized OS image installed on our cluster nodes. If the MachineConfigs for the `layered` MachineConfigPool are changed or the `Containerfile` is changed, a new `MachineOSBuild` will be created, the build will automatically start, and the image will be rolled out automatically to all of the nodes within the `layered` MachineConfigPool.

Reviewer: > the `Containerfile` is changed
Or more generically if the `MachineOSConfig` is changed?

Author: Great point!

# On-Cluster Layering Quickstart Guide

## Prerequisites
This quick-start guide assumes you have the following:

> - Access to an OpenShift 4.16+ cluster with TechPreview mode enabled.

Reviewer 2: Being picky to keep the working consistent with the end user docs. 
```suggestion
- Access to an OpenShift 4.16+ cluster with TechPreview feature gate enabled.
```
Do we need a link to [Enabling features using feature gates](https://docs.openshift.com/container-platform/4.16/nodes/clusters/nodes-cluster-enabling-features.html#nodes-cluster-enabling).

  # image to be deployed to.
  machineConfigPool:
    name: layered
  buildInputs:
    containerFile:
    # Here is where you can set the Containerfile for your MachineConfigPool.
    # You'll note that you need to provide an architecture. This is because this
    # will eventually support multiarch clusters. For now, only noArch is
    # supported.
    - containerfileArch: noarch
      content: |-
        <containerfile contents>
    # Here is where you can select an image builder type. For now, we only
    # support the "PodImageBuilder" type that we maintain ourselves. Future
    # integrations can / will include other build system integrations.
    imageBuilder:
      imageBuilderType: PodImageBuilder
    # The Machine OS Builder needs to know what pull secret it can use to pull
    # the base OS image.
    baseImagePullSecret:
      name: <secret-name>
    # Here is where you specify the name of the push secret you use to push
    # your newly-built image to.
    renderedImagePushSecret:
      name: <secret-name>
    # Here is where you specify the image registry to push your newly-built
    # images to.
    renderedImagePushspec: <final image pullspec>
  buildOutputs:

>     # Here is where you specify what image will be used on all of your nodes to

Reviewer 2: Is this right?
```suggestion
    # Here is where you specify what secret will be used on all of your nodes to
```

  --from-file=entitlement.pem=<(oc get secret/etc-pki-entitlement -n openshift-config-managed -o go-template='{{index .data "entitlement.pem" | base64decode }}') \
  --from-file=entitlement-key.pem=<(oc get secret/etc-pki-entitlement -n openshift-config-managed -o go-template='{{index .data "entitlement-key.pem" | base64decode }}')

# (optional): Create an ImageStream which will act as our image registry. This
# step can be omitted if you are using an external image registry such as
# Quay.io, though you will need to create the push and pull secrets in Quay.io
# as well as inside your cluster.
oc create imagestream os-images -n openshift-machine-config-operator

# (optional): To get the image registry pullspec for your ImageStream, you can run the following command:
# This will return "image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/os-images"
oc get imagestream/os-images  -n openshift-machine-config-operator -o=jsonpath='{.status.dockerImageRepository}'

# (optional): Since this walk-through will use an ImageStream, we need to get the push and
# pull secret associated with the builder service account. For the sake of this

> # demonstration, lets assume that this command returns the name

Reviewer 2: > lets assume that this command returns the name
Which command?

> # demonstration, lets assume that this command returns the name

Reviewer 2: ```suggestion
# demonstration, let's assume that this command returns the name
```

# "builder-dockercfg-123".
#
# If using an external image registry, this step can be omitted, although you
# will still need to create the push and pull secrets in the MCO namespace.
oc get secrets -o name -n openshift-machine-config-operator -o=jsonpath='{.items[?(@.metadata.annotations.openshift\.io\/internal-registry-auth-token\.service-account=="builder")].metadata.name}'

# (required): Create the layered MachineConfigPool:
cat << EOF | oc create -f -
---
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfigPool
metadata:
  name: layered
spec:
  machineConfigSelector:
    matchExpressions:
      - key: machineconfiguration.openshift.io/role
        operator: In
        values:
        - worker
        - layered
  nodeSelector:
    matchLabels:
      node-role.kubernetes.io/layered: ""
EOF
```

### Customized OS Image

> We will use the following Containerfile to install some useful packages on our cluster nodes:

Reviewer 2: Do we want to highlight the tree package here, as we use it later to verify? I missed this detail and didn't know why we were searching for the tree package later.

```suggestion
We will use the following Containerfile to install some useful packages on our cluster nodes, including the tree package, which we will use to verify the install later:
```

```Dockerfile
FROM configs AS final
RUN rpm-ostree install tree && \
    ostree container commit
```

A couple of notes about this Containerfile:

- Multiple build stages are allowed.

> - One must use the configs image target (`FROM configs AS final`) to

Reviewer 2: ```suggestion
- You must use the configs image target (`FROM configs AS final`) to
```

# the image registry onto each node.
export pullSecretName="builder-dockercfg-123"

# This is the name of the secret that will be used to pull the base OS image to
# the build pod to be consumed during the build.
export baseImagePullSecretName="global-pull-secret-copy"

# This has the contents of the Containerfile.
export containerfileContents="$(cat Containerfile)"

# Notice that we added ":latest" onto our image registry pullspec. This tag will
# not actually be used.
export imageRegistryPullspec="image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/os-images:latest"

yq -i e '.spec.buildInputs.baseImagePullSecret.name = strenv(baseImagePullSecretName)' ./layered-machineosconfig.yaml
yq -i e '.spec.buildInputs.renderedImagePushSecret.name = strenv(pushSecretName)' ./layered-machineosconfig.yaml
yq -i e '.spec.buildOutputs.currentImagePullSecret.name = strenv(pullSecretName)' ./layered-machineosconfig.yaml
yq -i e '.spec.buildInputs.containerFile[0].content = strenv(containerfileContents)' ./layered-machineosconfig.yaml
yq -i e '.spec.buildInputs.renderedImagePushspec = strenv(imageRegistryPullspec)' ./layered-machineosconfig.yaml

> ```

Reviewer 2: Do I have this right? 
Under _Write the sample MachineOSConfig to a YAML file_, the user creates the YAML file exactly as shown, with the variables (ie name: <secret-name>)? Then use the envvars and yq commands to populate those fields in the YAML?
